dbms/src/DataTypes/SmallestJSON/SmallestJSONSerialization.cpp

```diff
@@ -0,0 +1,357 @@
+               NumberImpl(field_column, ext::bit_cast<UInt64>(value));
+    }
+
+    bool Int64(Int64 value)
+    {
+        IColumn * mark_column = current_struct_info->getOrCreateMarkColumn();
+        IColumn * field_column = current_struct_info->getOrCreateDataColumn(NUMBER_DATA_TYPE);
+
+        return NumberImpl<true>(mark_column, UInt8(TypeIndex::Int64), maybeFillingJSONTypeToMark(current_struct_info)) &&
+               NumberImpl(field_column, ext::bit_cast<UInt64>(value));
+    }
+
+    bool Uint64(UInt64 value)
+    {
+        IColumn * mark_column = current_struct_info->getOrCreateMarkColumn();
+        IColumn * field_column = current_struct_info->getOrCreateDataColumn(NUMBER_DATA_TYPE);
+
+        return NumberImpl<true>(mark_column, UInt8(TypeIndex::UInt64), maybeFillingJSONTypeToMark(current_struct_info)) &&
+               NumberImpl(field_column, UInt64(value));
+    }
+
+    bool Double(Float64 value)
+    {
+        IColumn * mark_column = current_struct_info->getOrCreateMarkColumn();
+        IColumn * field_column = current_struct_info->getOrCreateDataColumn(NUMBER_DATA_TYPE);
+
+        return NumberImpl<true>(mark_column, UInt8(TypeIndex::Float64), maybeFillingJSONTypeToMark(current_struct_info)) &&
+               NumberImpl(field_column, ext::bit_cast<UInt64>(value));
+    }
+
+    bool StringImpl(IColumn * column, const char * data, rapidjson::SizeType length)
+    {
+        if (ColumnString * column_string = static_cast<ColumnString *>(column))
+        {
+            ColumnString::Chars & column_string_data = column_string->getChars();
+            ColumnString::Offsets & column_offsets_data = column_string->getOffsets();
+
+            if (!column_offsets_data.size() && row_size)
+            {
+                column_string_data.resize_fill(row_size, 0);
+                for (size_t index = 0; index < row_size; ++index)
+                    column_offsets_data.push_back(index + 1);
+            }
+
+            column_string_data.insert(data, data + length);
+            column_string_data.push_back(0);
+            column_offsets_data.push_back(column_string_data.size());
+            return true;
+        }
+        else
+            throw Exception("It is bug", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    bool String(const char * data, rapidjson::SizeType length, bool /*copy*/)
+    {
+        IColumn * mark_column = current_struct_info->getOrCreateMarkColumn();
+        IColumn * field_column = current_struct_info->getOrCreateDataColumn(STRING_DATA_TYPE);
+
+        return NumberImpl<true>(mark_column, UInt8(TypeIndex::String), maybeFillingJSONTypeToMark(current_struct_info)) &&
+               StringImpl(field_column, data, length);
+    }
+
+    bool Key(const char * data, rapidjson::SizeType length, bool /*copy*/)
+    {
+        current_struct_info = parent_struct_info->getOrCreateChildren(StringRef(data, length));
```

Comment:
Looks like if you insert a lot of JSONs into the same column like this:
```
{a1, 1}
{a2, 2}
{a3, 3}
...
{a10000, 100000}
```
you'll get very slow insertion and a lot of almost zero internal columns. Am I right?

Reply:
Yes, so I plan there might be a BigJSON type, which uses BJSON to store data and that is always BigJSON when first written, and is automatically converted to SmallestJSON only when JSON is small(when the number of columns in JSON is less than a some Setting, It may only happen at merge.)

Of course, there's some workload, and I'm going to turn it on when Smalllest JSON is finished.

Reply:
Have you considered using other alternatives instead of BJSON?
Maybe [fleece](https://github.com/couchbaselabs/fleece) is faster because with this format it's not necessary to parse to the end of the input stream to extract data.

Reply:
Great, let me see, what other JSON binary format do you recommend?

Reply:
> Yes, so I plan there might be a BigJSON type, which uses BJSON to store data and that is always BigJSON when first written, and is automatically converted to SmallestJSON only when JSON is small(when the number of columns in JSON is less than a some Setting, It may only happen at merge.)

I think that's a good idea. And I suggest not to create a new data type and after merge still use the same data type (only change its internally). I mean inside the same column class `ColumnSmallestJSON` we can have a flag which specifies whether this is a `ColumnSmallestJSONStruct`-based column or some kind of binary JSON.

Reply:
PostgreSQL has a data type named [JSONB](https://www.postgresql.org/docs/current/datatype-json.html) and it seems to me your data type can look to user similar to that. So maybe it's better to rename `SmallestJSON` => `JSONB`. What do you think?

Reply:
I've renamed it, and in fact, that's what I expected form the beginning. But because of the SmallestJSON is much easier for the user to notice that it works with small JSON.

Reply:
> Great, let me see, what other JSON binary format do you recommend? Or other suggestions for this idea ?

`fleece` seems to be very fast, I mean we can navigate through a binary JSON very quickly, i.e. we can quickly extract data. For example, to extract a value by key there a hash-map can be used. [BJSON](http://bjson.org) and [UBJSON](http://ubjson.org/#data_format) don't seem to be the fastest because they require to scan an object and sometimes even array from the beginning to extract a specific element by its key or index.

Reply:
> I've renamed it, and in fact, that's what I expected form the beginning. But because of the SmallestJSON is much easier for the user to notice that it works with small JSON.

I meant if there would be an implementation where the same data type `JSONB` can work as `SmallestJSON` and `BigJSON` with automatic conversion between the internal representations on merge like as you said then it wouldn't be necessary to emphasize for the user that these JSONs should be small at all.